### ISSUE TITLE
False Positive UnusedVariable

### ISSUE THREAD

**Issue by tomtomau** [vimeo/psalm#1129]
Version: dev-master

I'm having lots of difficulty trying to create a reproducible case but this code errors:
```php
    private function configureS3ClientForSuccess(
        string $storedFilePath,
        string $originalFilename,
        string $contentDisposition
    ): string {
        $expected = 'https://my_bucket/link/to/file.txt';

        $object = [
            'Bucket'                     => $this->adapter->getBucket(),
            'Key'                        => $storedFilePath,
            'ResponseContentDisposition' => \sprintf(
                '%s; filename="%s"',
                $contentDisposition,
                $originalFilename
            ),
        ];

        $command = $this->createMock(CommandInterface::class);
        $this->s3Client->expects(static::once())
            ->method('getCommand')
            ->with(
                'GetObject',
                new ArraySubset($object, true)
            )
            ->willReturn($command)
        ;
```

With:
```
Variable $object is never referenced
UnusedVariable
```

But if I change $object to be the built ArraySubset then it's fine:
```php
    private function configureS3ClientForSuccess(
        string $storedFilePath,
        string $originalFilename,
        string $contentDisposition
    ): string {
        $expected = 'https://my_bucket/link/to/file.txt';

        $object = new ArraySubset(
                [
                    'Bucket'                     => $this->adapter->getBucket(),
                    'Key'                        => $storedFilePath,
                    'ResponseContentDisposition' => \sprintf(
                        '%s; filename="%s"',
                        $contentDisposition,
                        $originalFilename
                    ),
                ], true
            );

        $command = $this->createMock(CommandInterface::class);
        $this->s3Client->expects(static::once())
            ->method('getCommand')
            ->with('GetObject', $object)
            ->willReturn($command)
        ;
```

For reference, ArraySubset is a PHPUnit 6.5 class, but the constructor typing seems fairly sane:
```php
    /**
     * @param array|\Traversable $subset
     * @param bool               $strict Check for object identity
     */
    public function __construct($subset, $strict = false)
    {
        parent::__construct();

        $this->strict = $strict;
        $this->subset = $subset;
    }
```

As I mentioned, I'm having difficulty trying to create a reproducible example and I'm not even sure how I'd describe it.

**Comment by tomtomau** [comment]
I think this is an issue with mocks, I will try using the new psalm-phpunit plugin and report back.

**Comment by muglug** [comment]
Still, though, it should still work
